fgpyo/io/__init__.py

```diff
@@ -0,0 +1,198 @@
+    directory until one is found that exists and is writeable.
+
+    In each scenario if a writeable file/directory is found the path is returned if not an
+    AssertionError is raised.
+
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+
+    Example:
+    >>> assert_path_is_writeable(path = Path("example.txt"))
+    // Assuming example.txt exists and is writeable
+    >>> example.txt
+    """
+    # If file exists, it must be writeable
+    if path.is_file():
+        assert os.access(path, os.W_OK), f"File exists but is not writebale: {path}"
+        return path
+
+    # Else if file doesnt exist and parent_must_exist is True then check
+    # that path.absolute().parent exists, is a directory and is writeable
+    elif path.is_file() is False & parent_must_exist is True:
+        assert (
+            path.absolute().parent.exists()
+            & path.absolute().parent.is_dir()
+            & os.access(path.absolute().parent, os.W_OK)
+        ), f"File does not have a writeable parent directory: {path}"
+        return path.absolute().parent
+
+    # Else if file doesn't exist and parent_must_exist is False, test parent until
+    # you find the first extent path, and check that it is a directory and is writeable.
+    elif path.is_dir() & parent_must_exist is False:
+        for directory in list(path.parents):
+            try:
+                assert (
+                    directory.is_dir() & directory.exists() & os.access(directory, os.W_OK)
+                ), f"File does not have a writeable parent directory: {path}"
+                return directory
+            except AssertionError:
+                pass
+        raise AssertionError(f"No parent directories of the path provided are writeable: {path}")
+
+
+def to_reader(path: Path) -> Union[io.TextIOWrapper, TextIO, IO[Any]]:
+    """Opens a Path for reading.
+
+    Args:
+        path: Path to read from
+
+    Example:
+    >>> reader = io.to_reader(path = Path("reader.txt"))
+    >>> reader.readlines()
+    >>> reader.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="rb")), encoding="utf-8")
+    else:
+        return path.open(mode="r")
+
+
+def to_writer(path: Path) -> Union[IO[Any], io.TextIOWrapper]:
+    """Opens a Path for reading and based on extension uses open() or gzip.open()
+
+    Args:
+        path: Path to write to
+
+    Example:
+    >>> writer = io.to_writer(path = Path("writer.txt"))
+    >>> writer.write(f'{something}\n')
+    >>> writer.close()
+    """
+    if path.suffix in COMPRESSED_FILE_EXTENSIONS:
+        return io.TextIOWrapper(cast(IO[bytes], gzip.open(path, mode="wb")), encoding="utf-8")
+    else:
+        return path.open(mode="w")
+
+
+def read_lines(path: Path, strip: bool = True) -> Union[Iterator[str], Generator[str, None, None]]:
```

Comment:
I think `strip` should be `False` by default as it will be less surprising.

Reply:
If you set the return type to just `Iterator[str]` does mypy complain?  Typing docs say that should work.